### PR TITLE
CP-44637: Update data model of Tunnel for cross-server private network

### PR DESCRIPTION
**Pull request by duobei**
1. a new field tunnel_id of type int to record the id for all the tunnels in this cross-server private network
2. `vxlan_mesh` added in tunnel_protocol to create the tunnel for cross-server private network

### CODE REVIEW

File: ocaml/idl/datamodel.ml [hunk 1]
```diff
@@ -2273,7 +2279,14 @@ module Tunnel = struct
...
         ; field ~ty:tunnel_protocol ~default_value:(Some (VEnum "gre"))
             ~lifecycle:[(Published, "1.250.0", "Add protocol field to tunnel")]
-            "protocol" "The protocol used for tunneling (either GRE or VxLAN)"
+            "protocol"
+            "The protocol used for tunneling (GRE, VxLAN, VxLAN_Mesh)"
+        ; field ~qualifier:StaticRO ~ty:Int ~default_value:(Some (VInt 0L))
+            ~lifecycle:
+              [(Published, rel_next, "VXLAN network identifier of tunnel")]
+            "tunnel_id"
+            "VXLAN network identifier of tunnel used in a cross-server private \
```
Comment (gangj) [inline ocaml/idl/datamodel.ml]: Removed the `VXLAN` here, I think otherwise `tunnel_id` -> `vxlan_tunnel_id`.

File: ocaml/xapi/nm.ml [hunk 2]
```diff
@@ -787,7 +787,7 @@ let bring_pif_up ~__context ?(management_interface = false) (pif : API.ref_PIF)
                 Db.Tunnel.get_protocol ~__context ~self:tunnel_ref
               in
               match protocol with
-              | `vxlan ->
+              | `vxlan | `vxlan_mesh ->
                   debug
                     "Opening VxLAN UDP port for tunnel with protocol 'vxlan'" ;
```
Comment (gangj) [inline ocaml/xapi/nm.ml]: Also update the log here would be better.

File: ocaml/xapi/nm.ml [hunk 3]
```diff
@@ -856,7 +856,7 @@ let bring_pif_down ~__context ?(force = false) (pif : API.ref_PIF) =
                        !Xapi_globs.firewall_port_config_script
                        ["close"; "4789"; "udp"]
                 )
-            | `gre ->
+            | `gre | `vxlan_mesh ->
```
Comment (gangj) [inline ocaml/xapi/nm.ml]: UDP port 4789 is opened for \`vxlan_mesh when `bring_pif_up`, why not close it when not used?

File: ocaml/xapi/xapi_pool.ml [hunk 4]
```diff
@@ -410,7 +410,11 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
   in
   (* Allow pool-join if host does not have any tunnels *)
   let assert_no_tunnels_on_me () =
-    if Db.Tunnel.get_all ~__context <> [] then (
+    let tunnels =
+      Db.Tunnel.get_refs_where ~__context
+        ~expr:(Not (Eq (Field "protocol", Literal "vxlan_mesh")))
```
Comment (gangj) [inline ocaml/xapi/xapi_pool.ml]: Why a tunnel for "private network" makes difference here?

File: ocaml/idl/datamodel.ml [hunk 5]
```diff
@@ -2187,7 +2187,13 @@ end
 module Tunnel = struct
   let tunnel_protocol =
     Enum
-      ("tunnel_protocol", [("gre", "GRE protocol"); ("vxlan", "VxLAN Protocol")])
+      ( "tunnel_protocol"
+      , [
+          ("gre", "GRE protocol")
+        ; ("vxlan", "VxLAN Protocol")
+        ; ("vxlan_mesh", "VxLAN Protocol for cross-server private network")
```
Comment (gangj) [inline ocaml/idl/datamodel.ml]: A detailed description for the "mesh" would be better, like "setup vxlan tunnels to every other host in the pool"

File: ocaml/idl/datamodel.ml [hunk 6]
```diff
@@ -2187,7 +2187,13 @@ end
 module Tunnel = struct
   let tunnel_protocol =
```
Comment (gangj) [inline ocaml/idl/datamodel.ml:2199]: As we had discussed, `vxlan_mesh` added into `tunnel_protocol` field is a little confusing, or we can add some comments here to clarify.

File: ocaml/xapi/xapi_pool.ml (hunk 4 above)
Comment (duobei) [reply ocaml/xapi/xapi_pool.ml]: Yeah, there should no difference here. A cross-server private network is limited in a pool.

File: ocaml/xapi/nm.ml (hunk 3 above)
Comment (duobei) [reply ocaml/xapi/nm.ml]: > If last tunnel with VxLAN protocol is unplugged, close VxLAN UDP port.

Ok, we should close the port if it's last tunnel with VxLAN protocol.

File: ocaml/idl/datamodel.ml [hunk 7]
```diff
@@ -2185,9 +2185,21 @@ module VLAN = struct
 end
 
 module Tunnel = struct
+  (* Adding vxlan_mesh in tunnel_protocol to distinguish cross-server private network maybe not reasonable.
+     I haven't found a better way yet. If there will be a better way found, it will be refactored.
```
Comment (gangj) [inline ocaml/idl/datamodel.ml:2189]: I would suggest to describe why we made this design decision in the comment.

**Review by gangj — APPROVED** [review]

**Pull request merged**